src/main/java/fi/aalto/cs/apluscourses/intellij/model/IntelliJCourseFactory.java

```diff
@@ -0,0 +1,33 @@
+package fi.aalto.cs.apluscourses.intellij.model;
+
+import com.intellij.openapi.project.Project;
+import fi.aalto.cs.apluscourses.model.Course;
+import fi.aalto.cs.apluscourses.model.CourseFactory;
+import fi.aalto.cs.apluscourses.model.Module;
+import java.net.URL;
+import java.util.List;
+import java.util.Map;
+import org.jetbrains.annotations.NotNull;
+
+public class IntelliJCourseFactory implements CourseFactory {
+
+  @NotNull
+  private final Project project;
+
+  public IntelliJCourseFactory(@NotNull Project project) {
+    this.project = project;
+  }
+
+  @Override
+  public Course createCourse(@NotNull String name,
+                             @NotNull List<Module> modules,
+                             @NotNull Map<String, String> requiredPlugins) {
+    return new IntelliJCourse(name, modules, requiredPlugins, project);
+  }
+
+  @Override
+  public Module createModule(@NotNull String name, @NotNull URL url) {
```

Comment:
to consider: does it make sense to create a module that is a part of the **project**, but is not tied to a **course**? In other words: [Project] 1 - * [Course] - * [Module]?

Reply:
I see your point. The problem is the order in which objects are created. Current course config file parser creates first `Module` objects and gives a list of them to the constructor of `Course`. Thus, `Module`s don't have links to `Course`. (We could add that link so that the `Course` would register itself calling `aModule.setCourse(this)` or something a like but currently there is no need for that -- besides this issue.) On the other hand, `Project` as a class of IntelliJ SDK is of course available already when `Module`s are created, so it is easy to give `Module` a link to `Project`.